### DISCUSSION TITLE
is it possible to do using pb_hook that check  a data  already exist?

### DISCUSSION THREAD

**Discussion by apujayed** [pocketbase/pocketbase#3553] [Q&A]
is it possible to do using onModelBeforeCreate that i have a filed called name if name MR A already exist and someone try to create same name MR A than throw an error

**Comment by ganigeorgiev** [comment]
You can do that with the hooks but if you need only a unique constraint you can create a "Unique index" from the Admin UI that will perform the validation for you:

![unique_screenshot](https://i.imgur.com/UVyo0ef.png)

**Reply by apujayed** [reply]
same thing if i want to do using pb_hooks only when create than how?

**Reply by ganigeorgiev** [reply]
You can query your database using the record dao helpers - https://pocketbase.io/docs/js-records/.
Here is one example:
```js
onModelBeforeCreate((e) => {
    const currentName = e.model.get("name");
    
    let hasRecord = false;

    try {
        // or you can also use directly the $app.dao().isRecordValueUnique("yourCollection", "name", currentName) bool check
        hasRecord = !!$app.dao().findFirstRecordByData("yourCollection", "name", currentName)
    } catch(_){}

    if (hasRecord) {
        throw new Error("name already exists")
    }
}, "yourCollection")
```

**Reply by ganigeorgiev** [reply]
If you want to return a formatted API error you can attach to the corresponding request hook instead (_in the future it will be possible to propagate API errors from other places_):

```js
// fired on create API request
onRecordBeforeCreateRequest((e) => {
    const currentName = e.record.get("name");

    let hasRecord = false;

    try {
        // or you can also use directly the $app.dao().isRecordValueUnique("yourCollection", "name", currentName) bool check
        hasRecord = !!$app.dao().findFirstRecordByData("yourCollection", "name", currentName)
    } catch(_){}

    if (hasRecord) {
        throw new BadRequestError("Failed to create the record.", {
            "name": new ValidationError("duplicated_name", "Name already exists."),
        })
    }
}, "yourCollection")
```

**Comment by apujayed** [comment]
duplicate error is working but if change the name and trying to re create it show an error that "The requested resource wasn't found."

```
onRecordBeforeCreateRequest((e) => {
   const currentName = e.record.get("name");

   if ($app.dao().findFirstRecordByData("userinfo", "name", currentName)) {
       throw new BadRequestError("Failed to create the record.", {
           "name": new ValidationError("duplicated_name", "Name already exists."),
       })
   }
}, "userinfo")

```

**Reply by ganigeorgiev** [reply]
Ah yes, sorry I forgot that the db call will throw an error if missing. I've updated the above examples.

**Reply by ganigeorgiev** [reply]
btw. if you want a unique constraint only for the create action, you can also create a partial unique index with something like:

```sql
CREATE UNIQUE INDEX `idx_duwsfGV` ON `userinfo` (`name`) WHERE `created` = `updated`
```
_(the `created` and `updated` fields are expected to be equal only on create)_
